### PR TITLE
Added weather features and test functions

### PR DESCRIPTION
**Pull request by ypang6**
Added CIWS EchoTop weather data parsers. Sample masked data for easy demonstration is also included.

### CODE REVIEW

**Comment by jmcfarland-swri** [comment]
Thank you for the contribution. I will review and provide some feedback.

In general, it is best to create a branch for your work instead of using the "master" branch. This way, you can pull changes from the upstream repository into your repository without worrying about conflicting with the work that you are doing on your branch. It also allows you to work on multiple things at once, each under a separate branch. If you want to move your work onto a branch, I can help you do this, but a new pull request will need to be submitted.

**Comment by jmcfarland-swri** [comment]
I'm going to go ahead and review, but I want to reiterate that you will be better off in the long run if you submit your pull request using a branch other than the master branch.  The reason I am pointing this out now is that changing branches will require a completely new pull request.

File: paraatm/io/weather_cube_ET.py [hunk 1]
```diff
@@ -0,0 +1,361 @@
...
+import os
+import time
+import math
+import datetime
+import numpy as np
+import pandas as pd
+from netCDF4 import Dataset
+
+
+class weather_cube_generator(object):
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:30]: Please use "CapWords" style for class names.  This is discussed in the "Style" section of the [Contributing Guide](https://github.com/ymlasu/para-atm/wiki/Contributing).

File: paraatm/io/weather_cube_ET.py [hunk 2]
```diff
@@ -0,0 +1,361 @@
...
+        y: int
+        The latitude index of the current locatuon in the grid map, after resize.
+        values: float
+        The weather grid array after resize.
+
+
+        Returns
+        -------
+        point_t_values: float
+        The averaged mean value of EchoTop around the current location.
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:90]: Missing indentation in the parameter descriptions.

File: paraatm/io/weather_cube_ET.py [hunk 3]
```diff
@@ -0,0 +1,361 @@
...
+    def get_cube(self, save_area=False, save_point=False):
+        """
+        Calculate the weather_tensor and the weather_point around the current location
+
+        Arguments
+        ---------
+        save_area: Boolean
+        If set to True, save the weather_area tensor into a .npy file. Default is False.
+        save_point: Boolean
+        If set to True, save the weather_point tensor into a .npy file. Default is False.
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:114]: Missing indentation for parameter descriptions.

File: paraatm/io/weather_cube_ET.py [hunk 4]
```diff
@@ -0,0 +1,361 @@
...
+
+    Parameters
+    ----------
+    array: float, int
+    The reference array.
+
+    Returns
+    -------
+    num: float
+    The number we are interested.
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:273]: Missing indentation in parameter descriptions.

File: paraatm/io/weather_cube_ET.py [hunk 5]
```diff
@@ -0,0 +1,361 @@
...
+    ----------
+    weather_path: str
+    The path to the weather data.
+    unix_time: int
+    The unix time of the current location
+
+    Returns
+    -------
+    file_path: str
+    The path to the closest ET weather file at the current location and the current time
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:293]: Missing indentation in parameter descriptions.

File: paraatm/io/weather_cube_ET.py [hunk 6]
```diff
@@ -0,0 +1,361 @@
...
+
+    Parameters
+    ----------
+    num: float
+    A four digit number represents the minute and second time.
+
+    Returns
+    -------
+    num: float
+    The cleared last four digits time
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:324]: Missing indentation in parameter descriptions.

File: paraatm/io/weather_cube_ET.py [hunk 7]
```diff
@@ -0,0 +1,361 @@
...
+
+    Parameters
+    ----------
+    str: str
+    The objective string of numbers to makeup
+
+    Returns
+    -------
+    str: str
+    The string with four numbers
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:348]: Missing indentation in parameter descriptions.

File: setup.py [hunk 8]
```diff
@@ -14,7 +14,9 @@
             'matplotlib',
             'jpype1==0.6.3',
             'scipy',
-            'sklearn'],
+            'sklearn',
+            'netCDF4',
+            'datetime'],
```
Comment (jmcfarland-swri) [inline setup.py:19]: `datetime` is built-in, so it does not need to be listed here as a dependency

File: paraatm/io/weather_cube_ET.py [hunk 9]
```diff
@@ -0,0 +1,361 @@
...
+    nearest_value = int(find_nearest_value(array, 0.001+np.asarray([int(eliminate_zeros(pin[-4:]))])))  # find the closest time for downloading data from CIWS
+    nearest_value = make_up_zeros(str(nearest_value))  # make up zeros for 0 230 500 730
+
+    filename = pin[:8] + "ET/ciws.EchoTop." + pin[:8] + "T" + str(pin[-6:-4]) + nearest_value + "Z.nc"
+
+    file_path = weather_path + filename
+    return file_path
+
+
+def eliminate_zeros(num):
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:312]: I suspect that maybe this function and others in your module might be intended only for internal use by the module, as opposed to being called directly by users of your module.  If that is the case, a good practice is to indicate this by prefixing the "private" functions with an underscore, e.g. `_eliminate_zeros`.  This helps make it clear which functions in your module people would want to use.

File: paraatm/io/weather_cube_ET.py [hunk 10]
```diff
@@ -0,0 +1,361 @@
...
+    """
+    clear the zeros in the output of the unix to time transformation.
+
+    Parameters
+    ----------
+    num: float
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:318]: This argument is a `str`, not a `float`, right?

File: paraatm/io/weather_cube_ET.py [hunk 11]
```diff
@@ -0,0 +1,361 @@
...
+    A four digit number represents the minute and second time.
+
+    Returns
+    -------
+    num: float
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:323]: Same as above, isn't this really a `str`?

File: paraatm/io/weather_cube_ET.py [hunk 12]
```diff
@@ -0,0 +1,361 @@
...
+    The string with four numbers
+    """
+    if len(str) == 4:
+        return str
+    if len(str) == 3:
+        return "0" + str
+    if len(str) == 2:
+        return "00" + str
+    if len(str) == 1:
+        return "000" + str
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:357]: How about just:
```suggestion
    '{:0>4}'.format(str)
```

File: paraatm/io/weather_cube_ET.py [hunk 13]
```diff
@@ -0,0 +1,361 @@
...
+    -------
+    num: float
+    The cleared last four digits time
+    """
+    if num[0] == '0' and num[1] == '0' and num[2] == '0':
+        return num[3]
+    if num[0] == '0' and num[1] == '0' and num[2] != '0':
+        return num[2:]
+    if num[0] == '0' and num[1] != '0':
+        return num[1:]
```
Comment (jmcfarland-swri) [inline paraatm/io/weather_cube_ET.py:331]: How about just:
```suggestion
    return num.lstrip('0')
```

File: paraatm/tests/test_weather_parser.py [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+
+# set parameters to the function in a dictionary
+cfg = {'cube_size': 20, 'resize_ratio': 1, 'downsample_ratio': 1}
+
+# Get the current directory
+cwd = os.path.dirname(os.getcwd())
+
+# Directory to weather data path
+cfg['weather_path'] = os.path.join(cwd, 'sample_data/weather_parser/echotop_data/')
+cfg['trajectory_path'] = os.path.join(cwd, 'sample_data/weather_parser/Weather_demo_track.csv')
```
Comment (jmcfarland-swri) [inline paraatm/tests/test_weather_parser.py:17]: This won't actually work when your test is called from anywhere other than the `paraatm` directory.  The best way is to get the location of the current module, and then find the `sample_data` directory relative to that.  Please see how this is done in [test_para_atm.py](../blob/master/paraatm/tests/test_para_atm.py).

File: paraatm/io/weather_cube_ET.py (hunk 1 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:30]: Thanks. I will modify the class name.

File: setup.py (hunk 8 above)
Comment (ypang6) [reply setup.py:19]: Thanks for the reminder. I didn't notice this before.

File: paraatm/io/weather_cube_ET.py (hunk 9 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:312]: Thanks for the suggestion. Yes, you are right. This is the internal function used only by the module. I will change it to private functions as your suggestion.

File: paraatm/io/weather_cube_ET.py (hunk 12 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:357]: This is really a good idea. Thanks for making this such clear. Many thanks.

File: paraatm/io/weather_cube_ET.py (hunk 13 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:331]: Many thanks.

File: paraatm/tests/test_weather_parser.py (hunk 14 above)
Comment (ypang6) [reply paraatm/tests/test_weather_parser.py:17]: That's a good point. This code assumes the test function stays in the /tests directory. I will change it to your approach.

**Comment by ypang6** [comment]
> I'm going to go ahead and review, but I want to reiterate that you will be better off in the long run if you submit your pull request using a branch other than the master branch. The reason I am pointing this out now is that changing branches will require a completely new pull request.

Thanks. I will submit it to my own branch instead.

File: paraatm/io/weather_cube_ET.py (hunk 7 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:348]: Thanks for pointing out the indentation issue.

File: paraatm/io/weather_cube_ET.py (hunk 11 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:323]: Yes, this is str. Sorry for the misleading information.

File: paraatm/io/weather_cube_ET.py (hunk 10 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:318]: Yes, this is str. Sorry for the misleading information.

File: paraatm/io/weather_cube_ET.py (hunk 12 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:357]: Many thanks again. This helped.

File: paraatm/io/weather_cube_ET.py (hunk 13 above)
Comment (ypang6) [reply paraatm/io/weather_cube_ET.py:331]: Many thanks for this lstrip method.

**Comment by jmcfarland-swri** [comment]
Closing, superseded by PR #22.

**Pull request closed without merging**